pyomo/contrib/sensitivity_toolbox/sens.py

```diff
@@ -7,216 +7,684 @@
-        preserve solver interface files
+    Parameters
+    ----------
+    model: Pyomo ConcreteModel
+        model should includes an objective function 
+    theta_names: list of strings
+        List of Var names
+    tee: bool, optional
+        Indicates that ef solver output should be teed
+    solver_options: dict, optional
+        Provides options to the solver (also the name of an attribute)
     
     Returns
     -------
-    model: ConcreteModel
-        The model modified for use with sipopt.  The returned model has
-        three :class:`Suffix` members defined:
-
-        - ``model.sol_state_1``: the approximated results at the
-          perturbation point
-        - ``model.sol_state_1_z_L``: the updated lower bound
-        - ``model.sol_state_1_z_U``: the updated upper bound
+    gradient_f: numpy.ndarray
+        gradient vector of the objective function with respect to the (decision variables, parameters) at the optimal solution
+    gradient_f_dic: dic
+        gradient_f with variable name as key 
+        e.g) dic = {'d(f)/d(x1)': 10.0, 'd(f)/d(x2)': 50.0, 'd(f)/d(p1)': 15.0, 'd(f)/d(p2)': 35.0}
+    gradient_c: numpy.ndarray
+        gradient vector of the constraints with respect to the (decision variables, parameters) at the optimal solution
+        Each row contains column number, row number, and value
+        If no constraint exists, return []
+    gradient_c: dic
+        gradient_c with constraint number and variable name as key 
+        e.g) dic = {'d(c1)/d(x1)': 1.0, 'd(c1)/d(p1)': -1.0, 'd(c2)/d(x2)': 1.0, 'd(c2)/d(p2)': -1.0}
+        Only non-zero gradients are included.
+    line_dic: dict
+        column numbers of the theta_names in the model. Index starts from 1
 
     Raises
     ------
-    ValueError
-        perturbList argument is expecting a List of Params
-    ValueError
-        length(paramSubList) must equal length(perturbList)
-    ValueError
-        paramSubList will not map to perturbList
-
+    RuntimeError
+        When ipopt or kaug or dotsens is not available
+    Exception
+        When ipopt fails 
     """
+    #Create the solver plugin using the ASL interface
+    ipopt = SolverFactory('ipopt',solver_io='nl')
+    if solver_options is not None:
+        ipopt.options = solver_options
+    kaug = SolverFactory('k_aug',solver_io='nl')
+    dotsens = SolverFactory('dot_sens',solver_io='nl')
+    if not ipopt.available(False):
+        raise RuntimeError('ipopt is not available')
+    if not kaug.available(False):
+        raise RuntimeError('k_aug is not available')
+    if not dotsens.available(False):
+        raise RuntimeError('dotsens is not available')
+
+    # Declare Suffixes
+    model.dual = Suffix(direction = Suffix.IMPORT)
+    model.ipopt_zL_out = Suffix(direction=Suffix.IMPORT)
+    model.ipopt_zU_out = Suffix(direction=Suffix.IMPORT)
+    model.ipopt_zL_in = Suffix(direction=Suffix.EXPORT)
+    model.ipopt_zU_in = Suffix(direction=Suffix.EXPORT)
```

Comment:
Should there be a check in case these suffixes already exist? This can probably just use the `_add_sensitivity_suffixes` function defined in this file. Or you can implement a similar `_add_kaug_suffixes` function if you don't want the sipopt suffixes.

Reply:
updated wiith   ` _add_sensitivity_suffixes(model)`